From 9dd8660bb10524ccbddd240c1c954f7732ce43eb Mon Sep 17 00:00:00 2001
From: Ma Wupeng <mawupeng1@huawei.com>
Date: Thu, 18 May 2023 11:21:30 +0800
Subject: [PATCH 032/263] [Huawei] mm/dpool: Stop counting free posioned pages

Offering: HULK
hulk inclusion
category: feature
bugzilla: 188703

-------------------------------

During dpool offline, only used pages is checked to determine whether this
dpool can be offline or not. So stop counting free posioned pages to make
this balance.

Signed-off-by: Ma Wupeng <mawupeng1@huawei.com>
---
 mm/dynamic_hugetlb.c | 17 ++++++++++-------
 1 file changed, 10 insertions(+), 7 deletions(-)

diff --git a/mm/dynamic_hugetlb.c b/mm/dynamic_hugetlb.c
index ca682a121485..7bdc486682f3 100644
--- a/mm/dynamic_hugetlb.c
+++ b/mm/dynamic_hugetlb.c
@@ -1314,17 +1314,20 @@ static int migrate_used_pages(struct dhugetlb_pool *hpool)
 		for (pfn = range->start; pfn <= range->end; pfn++) {
 			struct page *page = pfn_to_page(pfn);
 
-			if (PageHWPoison(page)) {
-				hpool->nr_poisoned_pages++;
+			/* Unlock and try migration. */
+			spin_unlock(&hpool->lock);
+			cond_resched();
+
+			if (!PagePool(page)) {
+				spin_lock(&hpool->lock);
 				continue;
 			}
 
-			/* Unlock and try migration. */
-			spin_unlock(&hpool->lock);
-			if (page->mapping != NULL)
-				ret |= do_migrate_range(pfn, pfn+1);
+			if (PageHWPoison(page))
+				/* Count used poisoned pages */
+				hpool->nr_poisoned_pages++;
 
-			cond_resched();
+			ret |= do_migrate_range(pfn, pfn+1);
 
 			spin_lock(&hpool->lock);
 		}
-- 
2.17.1

